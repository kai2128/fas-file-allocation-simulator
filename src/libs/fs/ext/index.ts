import { disk_block_size } from '~/libs/volume'

export interface ISuperblock {
  num_inodes: number
  num_blocks: number
  size_blocks: number
}

export interface IInode {
  size: number
  name: string
}

export interface IDiskBlock {
  next_block_num: number
  data: string
}

let sb: ISuperblock
const inodes: IInode[] = []
const dbs: IDiskBlock[] = []

// initialize fs
function create_fs() {
  sb.num_inodes = 10
  sb.num_blocks = 10
  sb.size_blocks = disk_block_size

  inodes.length = sb.num_inodes
  inodes.forEach((inode) => {
    inode.size = -1
    inode.name = ''
  })

  dbs.length = disk_block_size * sb.num_blocks
  for (let i = 0; i < sb.num_blocks; i++) {
    dbs[i].next_block_num = -1
  }


}

// load fs
function mount_fs() {

}

// write fs
function sync_fs() {

}
