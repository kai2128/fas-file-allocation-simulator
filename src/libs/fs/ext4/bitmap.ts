export class Bitmap {
  state: 0 | 1
  get used() {
    return this.state === 1
  }

  get free() {
    return this.state === 0
  }

  setUsed() {
    this.state = 1
  }

  setFree() {
    this.state = 0
  }

  constructor(
    public index: number,
  ) {
    this.state = 0
  }
}
