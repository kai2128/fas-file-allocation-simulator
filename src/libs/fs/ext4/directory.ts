import { remove } from 'lodash-es'
import type { Inode } from './inode'

interface File {
  name: string
  inode: number
}

export class Directory {
  hasFile(fileName: string) {
    return this.files.find(file => file.name === fileName)
  }

  inode: number
  files: File[]
  name: string

  constructor(inode: number, name: string) {
    this.name = name
    this.inode = inode
    this.files = []
  }

  addFile(inode: Inode) {
    this.files.push({
      name: inode.name,
      inode: inode.index,
    })
  }

  getFile(fileName: string) {
    return this.files.find(file => file.name === fileName)
  }

  deleteFile(fileName: string) {
    remove(this.files, { name: fileName })
  }
}
