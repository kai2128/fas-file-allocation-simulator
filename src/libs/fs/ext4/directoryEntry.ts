export class DirectoryEntry {
  constructor(
    public inode: number,
    public name: string,
  ) {
  }
}
