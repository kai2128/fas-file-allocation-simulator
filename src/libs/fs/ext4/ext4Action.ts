import type { FSActions } from '../types'

export const ext4Action: FSActions = {
  create: {
    steps: [
      {
        index: 0,
        description: 'Check if file with same name exists',
      },
      {
        index: 1,
        description: 'Allocate free inode for the file',
      },
      {
        index: 2,
        description: `
        Search for free blocks from block bitmaps for allocation.
        `,
      },
      {
        index: 3,
        description: `
        For each contiguous free blocks found, create an extent for 
        these blocks by recording its starting block and its length.
        `,
      },
      {
        index: 4,
        description: `
        Based on extents, set blocks as used in block bitmap.`,
      },
      {
        index: 5,
        description: `
        Add file to root directory.`,
      },
    ],
  },
  delete: {
    steps: [
      {
        index: 0,
        description: `
        Search for the directory entry with same name in the root directory.
        `,
      },
      {
        index: 1,
        description: `
        From the directory entry found, retrieve the inode number. From inode number, retrieve inode data.
        `,
      },
      {
        index: 2,
        description: `
        Retrieve allocated blocks from all extents.`,
      },
      {
        index: 3,
        description: `
        Mark all allocated blocks recorded in extents as free in the block bitmap.
        `,
      },
      {
        index: 4,
        description: `
        Mark inode as free and remove inode from directory.
        `,
      },
    ],
  },
  append: {
    steps: [
      {
        index: 0,
        description: `
        Search for the directory entry with same name in the root directory.
        `,
      },
      {
        index: 1,
        description: `
        From the directory entry found, retrieve the inode number. From inode number, retrieve inode data.
        `,
      },
      {
        index: 2,
        description: `
        Allocated blocks are stored in extents. Retrieve allocated blocks from all extents.`,
      },
      {
        index: 3,
        description: `
        Locate last allocated block of the file from extent data.
        `,
      },
      {
        index: 4,
        description: `
        Start searching for free block from last allocated block. For each contiguous free blocks found, create an extent for these blocks by recording its starting block and its length.
        `,
      },
      {
        index: 5,
        description: `
        Merge all contiguous extents to one extents.
        `,
      },
      {
        index: 6,
        description: `
        Based on extents, set blocks as used in block bitmap.
        `,
      },
    ],
  },
  write: {
    steps: [
      {
        index: 0,
        description: `
        Search for the directory entry with same name in the root directory.
        `,
      },
      {
        index: 1,
        description: `
        From the directory entry found, retrieve the inode number. From inode number, retrieve inode data and allocated blocks.
        `,
      },
      {
        index: 2,
        description: `
        Delete previously allocated blocks.
        `,
      },
      {
        index: 3,
        description: `
        Start searching for free block from first allocated block. For each contiguous free blocks found, create an extent for these blocks by recording its starting block and its length.
        `,
      },
      {
        index: 4,
        description: `
        Merge all contiguous extents to one extents.
        `,
      },
      {
        index: 5,
        description: `
        Based on extents, set blocks as used in block bitmap.
        `,
      },
    ],
  },
  read: {
    steps: [
      {
        index: 0,
        description: `
        Search for the directory entry with same name in the root directory.
        `,
      },
      {
        index: 1,
        description: `
        From the directory entry found, retrieve the inode number. From inode number, retrieve inode data.
        `,
      },
      {
        index: 2,
        description: `
        Allocated blocks are stored in extents. Retrieve allocated blocks from all extents.`,
      },
    ],
  },
}
