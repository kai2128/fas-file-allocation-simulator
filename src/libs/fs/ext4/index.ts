export * from './inode'
export * from './directoryEntry'
export * from './directory'
export * from './ext4Action'
export * from './ext4Animation'
export * from './ext4'
export * from './extent'
export * from './inodeTable'
export * from './bitmap'
