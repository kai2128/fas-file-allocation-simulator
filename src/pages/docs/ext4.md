# ext4


<route lang="yaml">
meta:
  layout: docs
  head: FAS - ext4
</route>