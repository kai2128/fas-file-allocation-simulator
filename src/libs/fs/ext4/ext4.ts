import { cloneDeep, filter, last } from 'lodash-es'
import type { FSApi, FileReaded } from './../types'
import { Bitmap } from './bitmap'
import { Directory } from './directory'
import { Inode } from './inode'
import { InodeTable } from './inodeTable'
import type { Disk } from '~/libs/volume'
import { ERRCODE, FSError } from '~/libs/error/fserror'

export class Ext4 implements FSApi {
  name = 'ext4'
  inodeBitmap: Bitmap[]
  blockBitmap: Bitmap[]
  inodeTable: InodeTable
  rootDirectory: Directory
  disk: Disk

  constructor(disk: Disk) {
    this.disk = disk
    this.inodeBitmap = new Array(disk.disk_size).fill({}).map((v, i) => {
      return new Bitmap(i)
    })
    this.blockBitmap = new Array(disk.disk_size).fill({}).map((v, i) => {
      return new Bitmap(i)
    })
    this.inodeTable = new InodeTable(disk.disk_size)
    this.rootDirectory = this.createRootDirectory()
  }

  createRootDirectory(): Directory {
    const inodeNumber = this.getFreeInodeIndex().index
    const root = new Directory(inodeNumber, '/')
    const inode = new Inode(inodeNumber, '/', 1, this.blockBitmap, 'directory')
    this.updateInodeTable(inode)
    this.writeToDisk(inode)
    return root
  }

  private updateInodeTable(inode: Inode) {
    this.inodeTable.inodes[inode.index] = inode
    this.inodeBitmap[inode.index].setUsed()
  }

  private writeToDisk(inode: Inode): void {
    const blocks = inode.extentTree.getAllocatedBlocks()
    blocks.forEach((block) => {
      this.blockBitmap[block].setUsed()
      this.disk.setUsed(block, inode.color, inode)
    })
  }

  private getFreeInodeIndex() {
    return this.inodeBitmap.find(v => v.free)!
  }

  private checkSpace(size?: number) {
    const freeBlocks = filter(this.blockBitmap, bitmap => bitmap.free)
    if (freeBlocks.length < 1)
      throw new FSError(ERRCODE.ESPACE, 'no enough free block')

    if (size) {
      if (freeBlocks.length - size < 0)
        throw new FSError(ERRCODE.ESPACE, 'no enough free cluster')
    }
  }

  private checkUnique(fileName: string) {
    if (this.rootDirectory.hasFile(fileName))
      throw new FSError(ERRCODE.EEXIST, 'file already exists')
  }

  private checkExist(fileName: string) {
    if (!this.rootDirectory.hasFile(fileName))
      throw new FSError(ERRCODE.ENOENT, 'file not found')
  }

  private getInodeFromDirectory(fileName: string) {
    const inodeNumber = this.rootDirectory.getFile(fileName)!.inode
    return this.inodeTable.inodes[inodeNumber]
  }

  createFile(fileName: string, size: number): void {
    this.checkSpace(size)
    this.checkUnique(fileName)

    const inode = new Inode(this.getFreeInodeIndex().index, fileName, size, this.blockBitmap)
    this.updateInodeTable(inode)
    this.writeToDisk(inode)
    this.rootDirectory.addFile(inode)
  }

  deleteFile(fileName: string): void {
    this.checkExist(fileName)

    const inode = this.getInodeFromDirectory(fileName)
    this.rootDirectory.deleteFile(fileName)
    this.inodeBitmap[inode.index].setFree()
    this.inodeTable.inodes[inode.index].setFree()

    inode.setAllocatedBlockFree(this.blockBitmap, this.disk)
  }

  readFile(fileName: string): Inode {
    this.checkExist(fileName)
    const inode = this.getInodeFromDirectory(fileName)
    return inode
  }

  writeFile(fileName: string, size: number): void {
    this.checkExist(fileName)
    const inode = this.getInodeFromDirectory(fileName)

    if (size > inode.size) { // only check disk space if new size is larger than old size{
      this.checkSpace(size - inode.size)
    }

    this.deleteFile(fileName)
    this.createFile(fileName, size)
  }

  appendFile(fileName: string, size: number): void {
    this.checkExist(fileName)
    this.checkSpace(size)
    const inode = this.getInodeFromDirectory(fileName)
    inode.setSize(size)
    inode.extentTree.appendBlockToExtent(size, this.blockBitmap)
    this.writeToDisk(inode)
  }

  static format(disk: Disk) {
    const fs = new Ext4(disk)
    log(`Created disk with size of ${disk.total_units} and formatted with ${fs.name}`)
    return fs
  }

  fs_create(fileName: string, size: number): void {
    this.createFile(fileName, size)
    log(`File ${fileName} created with size ${size}.`)
  }

  fs_append(fileName: string, size: number): void {
    this.appendFile(fileName, size)
    log(`File ${fileName} appended with size ${size}.`)
  }

  fs_read(fileName: string): void {
    this.readFile(fileName)
    log(`File ${fileName} read.`)
  }

  fs_delete(fileName: string): void {
    this.deleteFile(fileName)
    log(`File ${fileName} deleted.`)
  }

  fs_write(fileName: string, size: number): void {
    this.writeFile(fileName, size)
    log(`File ${fileName} created with size ${size}.`)
  }

  fs_defragmentation(): void {
    this.rootDirectory.files.forEach((file) => {
      const inode = this.getInodeFromDirectory(file.name)
      inode.setAllocatedBlockFree(this.blockBitmap, this.disk)
    })

    this.rootDirectory.files.forEach((file) => {
      const inode = this.getInodeFromDirectory(file.name)
      inode.extentTree.allocateFreeBlockToExtent(this.blockBitmap)
      this.writeToDisk(inode)
    })
    log('Defragmentation done.')
  }

  get fs_files(): FileReaded[] {
    return this.rootDirectory.files.map((v) => {
      const inode = this.inodeTable.inodes[v.inode]
      return {
        data: {
          name: inode.name,
          size: inode.size,
          dateCreated: inode.dateCreated,
          inodeNumber: inode.index,
          color: inode.color,
          type: inode.type,
        },
      }
    })
  }

  clone(disk: Disk): FSApi {
    const cloneFs = new Ext4(disk)
    cloneFs.inodeBitmap = cloneDeep(this.inodeBitmap)
    cloneFs.blockBitmap = cloneDeep(this.blockBitmap)
    cloneFs.inodeTable = cloneDeep(this.inodeTable)
    cloneFs.rootDirectory = cloneDeep(this.rootDirectory)
    cloneFs.disk = disk
    cloneFs.name = this.name
    return cloneFs
  }
}
