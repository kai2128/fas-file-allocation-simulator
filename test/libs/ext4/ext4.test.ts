import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest'
import { Ext4 } from '~/libs/fs/ext4'
import { Disk } from '~/libs/volume'

function setupExt() {
  const disk = new Disk(20)
  return {
    disk,
    ext4: Ext4.format(disk),
  }
}

describe('ext4 test', () => {
  beforeEach(() => {
    vi.useFakeTimers()
    vi.setSystemTime(new Date('2020-01-01T00:00:00.000Z'))
  })

  afterEach(() => {
    vi.useRealTimers()
  })

  test('format correctly', () => {
    const { ext4 } = setupExt()
    expect(ext4.inodeTable).toMatchSnapshot()
    expect(ext4.rootDirectory).toMatchInlineSnapshot(`
      Directory {
        "color": "#fcb874",
        "files": [],
        "inode": 0,
        "name": "/",
      }
    `)
    expect(ext4.inodeBitmap).toMatchSnapshot()
    expect(ext4.blockBitmap).toMatchSnapshot()
  })

  test('create file', () => {
    const { ext4, disk } = setupExt()
    ext4.createFile('test.txt', 2)
    expect(ext4.inodeTable.inodes[1].entry).toMatchSnapshot()
    expect(ext4.rootDirectory).toMatchSnapshot()
    expect(disk.read(1, 2)).toMatchSnapshot()
  })

  test('delete file', () => {
    const { ext4, disk } = setupExt()
    ext4.createFile('test.txt', 2)
    ext4.deleteFile('test.txt')
    expect(ext4.inodeTable.inodes[1].entry).toMatchSnapshot()
    expect(ext4.rootDirectory).toMatchSnapshot()
    expect(disk.read(1, 2)).toMatchSnapshot()
  })

  test('write file', () => {
    const { ext4, disk } = setupExt()
    ext4.createFile('test.txt', 2)
    ext4.writeFile('test.txt', 4)
    expect(ext4.inodeTable.inodes[1].entry).toMatchSnapshot()
    expect(ext4.rootDirectory).toMatchSnapshot()
    expect(disk.read(1, 4)).toMatchSnapshot()
  })

  test('append file', () => {
    const { ext4, disk } = setupExt()
    ext4.createFile('test.txt', 2)
    ext4.appendFile('test.txt', 4)
    expect(ext4.inodeTable.inodes[1].entry).toMatchSnapshot()
    expect(ext4.rootDirectory).toMatchSnapshot()
    expect(ext4.inodeTable.inodes[1].allocatedBlock).toMatchSnapshot()
    expect(disk.readUnit(1)).toMatchSnapshot()
    expect(disk.readUnit(6)).toMatchSnapshot()
    ext4.appendFile('test.txt', 1)
    expect(disk.readUnit(7)).toMatchSnapshot()
  })

  test('allocating file without contiguously', () => {
    const { ext4, disk } = setupExt()
    ext4.createFile('test.txt', 2)
    ext4.createFile('test2.txt', 2)
    ext4.createFile('test3.txt', 2)
    ext4.writeFile('test2.txt', 1)
    ext4.appendFile('test.txt', 3)
    expect(ext4.inodeTable.inodes[1].entry).toMatchSnapshot()
    expect(ext4.inodeTable.inodes[2].entry).toMatchSnapshot()
    expect(ext4.inodeTable.inodes[3].entry).toMatchSnapshot()
    expect(disk.read(1, 8)).toMatchSnapshot()
  })
})
