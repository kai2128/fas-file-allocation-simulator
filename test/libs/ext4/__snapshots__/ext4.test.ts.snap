// Vitest Snapshot v1

exports[`ext4 test > allocating file without contiguously 1`] = `
{
  "color": "#f477a3",
  "dateCreated": 1577836800000,
  "inode": 1,
  "name": "test.txt",
  "size": 5,
  "type": "file",
}
`;

exports[`ext4 test > allocating file without contiguously 2`] = `
{
  "color": "#f28ab3",
  "dateCreated": 1577836800000,
  "inode": 2,
  "name": "test2.txt",
  "size": 1,
  "type": "file",
}
`;

exports[`ext4 test > allocating file without contiguously 3`] = `
{
  "color": "#f296a4",
  "dateCreated": 1577836800000,
  "inode": 3,
  "name": "test3.txt",
  "size": 2,
  "type": "file",
}
`;

exports[`ext4 test > allocating file without contiguously 4`] = `
[
  {
    "offset": 1,
    "state": {
      "color": "#f477a3",
      "data": Inode {
        "color": "#f477a3",
        "dateCreated": 1577836800000,
        "extentTree": ExtentTree {
          "extents": [
            Extent {
              "length": 2,
              "start": 1,
            },
            Extent {
              "length": 1,
              "start": 4,
            },
            Extent {
              "length": 2,
              "start": 7,
            },
          ],
          "fileSize": 5,
        },
        "index": 1,
        "name": "test.txt",
        "size": 5,
        "state": "used",
        "type": "file",
      },
      "free": false,
      "reserved": false,
      "used": true,
    },
  },
  {
    "offset": 2,
    "state": {
      "color": "#f477a3",
      "data": Inode {
        "color": "#f477a3",
        "dateCreated": 1577836800000,
        "extentTree": ExtentTree {
          "extents": [
            Extent {
              "length": 2,
              "start": 1,
            },
            Extent {
              "length": 1,
              "start": 4,
            },
            Extent {
              "length": 2,
              "start": 7,
            },
          ],
          "fileSize": 5,
        },
        "index": 1,
        "name": "test.txt",
        "size": 5,
        "state": "used",
        "type": "file",
      },
      "free": false,
      "reserved": false,
      "used": true,
    },
  },
  {
    "offset": 3,
    "state": {
      "color": "#f28ab3",
      "data": Inode {
        "color": "#f28ab3",
        "dateCreated": 1577836800000,
        "extentTree": ExtentTree {
          "extents": [
            Extent {
              "length": 1,
              "start": 3,
            },
          ],
          "fileSize": 1,
        },
        "index": 2,
        "name": "test2.txt",
        "size": 1,
        "state": "used",
        "type": "file",
      },
      "free": false,
      "reserved": false,
      "used": true,
    },
  },
  {
    "offset": 4,
    "state": {
      "color": "#f477a3",
      "data": Inode {
        "color": "#f477a3",
        "dateCreated": 1577836800000,
        "extentTree": ExtentTree {
          "extents": [
            Extent {
              "length": 2,
              "start": 1,
            },
            Extent {
              "length": 1,
              "start": 4,
            },
            Extent {
              "length": 2,
              "start": 7,
            },
          ],
          "fileSize": 5,
        },
        "index": 1,
        "name": "test.txt",
        "size": 5,
        "state": "used",
        "type": "file",
      },
      "free": false,
      "reserved": false,
      "used": true,
    },
  },
  {
    "offset": 5,
    "state": {
      "color": "#f296a4",
      "data": Inode {
        "color": "#f296a4",
        "dateCreated": 1577836800000,
        "extentTree": ExtentTree {
          "extents": [
            Extent {
              "length": 2,
              "start": 5,
            },
          ],
          "fileSize": 2,
        },
        "index": 3,
        "name": "test3.txt",
        "size": 2,
        "state": "used",
        "type": "file",
      },
      "free": false,
      "reserved": false,
      "used": true,
    },
  },
  {
    "offset": 6,
    "state": {
      "color": "#f296a4",
      "data": Inode {
        "color": "#f296a4",
        "dateCreated": 1577836800000,
        "extentTree": ExtentTree {
          "extents": [
            Extent {
              "length": 2,
              "start": 5,
            },
          ],
          "fileSize": 2,
        },
        "index": 3,
        "name": "test3.txt",
        "size": 2,
        "state": "used",
        "type": "file",
      },
      "free": false,
      "reserved": false,
      "used": true,
    },
  },
  {
    "offset": 7,
    "state": {
      "color": "#f477a3",
      "data": Inode {
        "color": "#f477a3",
        "dateCreated": 1577836800000,
        "extentTree": ExtentTree {
          "extents": [
            Extent {
              "length": 2,
              "start": 1,
            },
            Extent {
              "length": 1,
              "start": 4,
            },
            Extent {
              "length": 2,
              "start": 7,
            },
          ],
          "fileSize": 5,
        },
        "index": 1,
        "name": "test.txt",
        "size": 5,
        "state": "used",
        "type": "file",
      },
      "free": false,
      "reserved": false,
      "used": true,
    },
  },
  {
    "offset": 8,
    "state": {
      "color": "#f477a3",
      "data": Inode {
        "color": "#f477a3",
        "dateCreated": 1577836800000,
        "extentTree": ExtentTree {
          "extents": [
            Extent {
              "length": 2,
              "start": 1,
            },
            Extent {
              "length": 1,
              "start": 4,
            },
            Extent {
              "length": 2,
              "start": 7,
            },
          ],
          "fileSize": 5,
        },
        "index": 1,
        "name": "test.txt",
        "size": 5,
        "state": "used",
        "type": "file",
      },
      "free": false,
      "reserved": false,
      "used": true,
    },
  },
]
`;

exports[`ext4 test > append file 1`] = `
{
  "color": "#f477a3",
  "dateCreated": 1577836800000,
  "inode": 1,
  "name": "test.txt",
  "size": 6,
  "type": "file",
}
`;

exports[`ext4 test > append file 2`] = `
Directory {
  "color": "#fcb874",
  "files": [
    {
      "inode": 1,
      "name": "test.txt",
    },
  ],
  "inode": 0,
  "name": "/",
}
`;

exports[`ext4 test > append file 3`] = `
[
  1,
  2,
  3,
  4,
  5,
  6,
]
`;

exports[`ext4 test > append file 4`] = `
{
  "offset": 1,
  "state": {
    "color": "#f477a3",
    "data": Inode {
      "color": "#f477a3",
      "dateCreated": 1577836800000,
      "extentTree": ExtentTree {
        "extents": [
          Extent {
            "length": 6,
            "start": 1,
          },
        ],
        "fileSize": 6,
      },
      "index": 1,
      "name": "test.txt",
      "size": 6,
      "state": "used",
      "type": "file",
    },
    "free": false,
    "reserved": false,
    "used": true,
  },
}
`;

exports[`ext4 test > append file 5`] = `
{
  "offset": 6,
  "state": {
    "color": "#f477a3",
    "data": Inode {
      "color": "#f477a3",
      "dateCreated": 1577836800000,
      "extentTree": ExtentTree {
        "extents": [
          Extent {
            "length": 6,
            "start": 1,
          },
        ],
        "fileSize": 6,
      },
      "index": 1,
      "name": "test.txt",
      "size": 6,
      "state": "used",
      "type": "file",
    },
    "free": false,
    "reserved": false,
    "used": true,
  },
}
`;

exports[`ext4 test > append file 6`] = `
{
  "offset": 7,
  "state": {
    "color": "#f477a3",
    "data": Inode {
      "color": "#f477a3",
      "dateCreated": 1577836800000,
      "extentTree": ExtentTree {
        "extents": [
          Extent {
            "length": 7,
            "start": 1,
          },
        ],
        "fileSize": 7,
      },
      "index": 1,
      "name": "test.txt",
      "size": 7,
      "state": "used",
      "type": "file",
    },
    "free": false,
    "reserved": false,
    "used": true,
  },
}
`;

exports[`ext4 test > create file 1`] = `
{
  "color": "#f477a3",
  "dateCreated": 1577836800000,
  "inode": 1,
  "name": "test.txt",
  "size": 2,
  "type": "file",
}
`;

exports[`ext4 test > create file 2`] = `
Directory {
  "color": "#fcb874",
  "files": [
    {
      "inode": 1,
      "name": "test.txt",
    },
  ],
  "inode": 0,
  "name": "/",
}
`;

exports[`ext4 test > create file 3`] = `
[
  {
    "offset": 1,
    "state": {
      "color": "#f477a3",
      "data": Inode {
        "color": "#f477a3",
        "dateCreated": 1577836800000,
        "extentTree": ExtentTree {
          "extents": [
            Extent {
              "length": 2,
              "start": 1,
            },
          ],
          "fileSize": 2,
        },
        "index": 1,
        "name": "test.txt",
        "size": 2,
        "state": "used",
        "type": "file",
      },
      "free": false,
      "reserved": false,
      "used": true,
    },
  },
  {
    "offset": 2,
    "state": {
      "color": "#f477a3",
      "data": Inode {
        "color": "#f477a3",
        "dateCreated": 1577836800000,
        "extentTree": ExtentTree {
          "extents": [
            Extent {
              "length": 2,
              "start": 1,
            },
          ],
          "fileSize": 2,
        },
        "index": 1,
        "name": "test.txt",
        "size": 2,
        "state": "used",
        "type": "file",
      },
      "free": false,
      "reserved": false,
      "used": true,
    },
  },
]
`;

exports[`ext4 test > delete file 1`] = `
{
  "color": "#f477a3",
  "dateCreated": 1577836800000,
  "inode": 1,
  "name": "DELETED",
  "size": 0,
  "type": "file",
}
`;

exports[`ext4 test > delete file 2`] = `
Directory {
  "color": "#fcb874",
  "files": [],
  "inode": 0,
  "name": "/",
}
`;

exports[`ext4 test > delete file 3`] = `
[
  {
    "offset": 1,
    "state": {
      "color": "#ffffff",
      "data": {},
      "free": true,
      "reserved": false,
      "used": false,
    },
  },
  {
    "offset": 2,
    "state": {
      "color": "#ffffff",
      "data": {},
      "free": true,
      "reserved": false,
      "used": false,
    },
  },
]
`;

exports[`ext4 test > format correctly 1`] = `
InodeTable {
  "inodes": [
    Inode {
      "color": "#fcb874",
      "dateCreated": 1577836800000,
      "extentTree": ExtentTree {
        "extents": [
          Extent {
            "length": 1,
            "start": 0,
          },
        ],
        "fileSize": 1,
      },
      "index": 0,
      "name": "/",
      "size": 1,
      "state": "used",
      "type": "directory",
    },
    ,
    ,
    ,
    ,
    ,
    ,
    ,
    ,
    ,
    ,
    ,
    ,
    ,
    ,
    ,
    ,
    ,
    ,
    ,
  ],
}
`;

exports[`ext4 test > format correctly 3`] = `
[
  Bitmap {
    "index": 0,
    "state": 1,
  },
  Bitmap {
    "index": 1,
    "state": 0,
  },
  Bitmap {
    "index": 2,
    "state": 0,
  },
  Bitmap {
    "index": 3,
    "state": 0,
  },
  Bitmap {
    "index": 4,
    "state": 0,
  },
  Bitmap {
    "index": 5,
    "state": 0,
  },
  Bitmap {
    "index": 6,
    "state": 0,
  },
  Bitmap {
    "index": 7,
    "state": 0,
  },
  Bitmap {
    "index": 8,
    "state": 0,
  },
  Bitmap {
    "index": 9,
    "state": 0,
  },
  Bitmap {
    "index": 10,
    "state": 0,
  },
  Bitmap {
    "index": 11,
    "state": 0,
  },
  Bitmap {
    "index": 12,
    "state": 0,
  },
  Bitmap {
    "index": 13,
    "state": 0,
  },
  Bitmap {
    "index": 14,
    "state": 0,
  },
  Bitmap {
    "index": 15,
    "state": 0,
  },
  Bitmap {
    "index": 16,
    "state": 0,
  },
  Bitmap {
    "index": 17,
    "state": 0,
  },
  Bitmap {
    "index": 18,
    "state": 0,
  },
  Bitmap {
    "index": 19,
    "state": 0,
  },
]
`;

exports[`ext4 test > format correctly 4`] = `
[
  Bitmap {
    "index": 0,
    "state": 1,
  },
  Bitmap {
    "index": 1,
    "state": 0,
  },
  Bitmap {
    "index": 2,
    "state": 0,
  },
  Bitmap {
    "index": 3,
    "state": 0,
  },
  Bitmap {
    "index": 4,
    "state": 0,
  },
  Bitmap {
    "index": 5,
    "state": 0,
  },
  Bitmap {
    "index": 6,
    "state": 0,
  },
  Bitmap {
    "index": 7,
    "state": 0,
  },
  Bitmap {
    "index": 8,
    "state": 0,
  },
  Bitmap {
    "index": 9,
    "state": 0,
  },
  Bitmap {
    "index": 10,
    "state": 0,
  },
  Bitmap {
    "index": 11,
    "state": 0,
  },
  Bitmap {
    "index": 12,
    "state": 0,
  },
  Bitmap {
    "index": 13,
    "state": 0,
  },
  Bitmap {
    "index": 14,
    "state": 0,
  },
  Bitmap {
    "index": 15,
    "state": 0,
  },
  Bitmap {
    "index": 16,
    "state": 0,
  },
  Bitmap {
    "index": 17,
    "state": 0,
  },
  Bitmap {
    "index": 18,
    "state": 0,
  },
  Bitmap {
    "index": 19,
    "state": 0,
  },
]
`;

exports[`ext4 test > write file 1`] = `
{
  "color": "#f477a3",
  "dateCreated": 1577836800000,
  "inode": 1,
  "name": "test.txt",
  "size": 4,
  "type": "file",
}
`;

exports[`ext4 test > write file 2`] = `
Directory {
  "color": "#fcb874",
  "files": [
    {
      "inode": 1,
      "name": "test.txt",
    },
  ],
  "inode": 0,
  "name": "/",
}
`;

exports[`ext4 test > write file 3`] = `
[
  {
    "offset": 1,
    "state": {
      "color": "#f477a3",
      "data": Inode {
        "color": "#f477a3",
        "dateCreated": 1577836800000,
        "extentTree": ExtentTree {
          "extents": [
            Extent {
              "length": 4,
              "start": 1,
            },
          ],
          "fileSize": 4,
        },
        "index": 1,
        "name": "test.txt",
        "size": 4,
        "state": "used",
        "type": "file",
      },
      "free": false,
      "reserved": false,
      "used": true,
    },
  },
  {
    "offset": 2,
    "state": {
      "color": "#f477a3",
      "data": Inode {
        "color": "#f477a3",
        "dateCreated": 1577836800000,
        "extentTree": ExtentTree {
          "extents": [
            Extent {
              "length": 4,
              "start": 1,
            },
          ],
          "fileSize": 4,
        },
        "index": 1,
        "name": "test.txt",
        "size": 4,
        "state": "used",
        "type": "file",
      },
      "free": false,
      "reserved": false,
      "used": true,
    },
  },
  {
    "offset": 3,
    "state": {
      "color": "#f477a3",
      "data": Inode {
        "color": "#f477a3",
        "dateCreated": 1577836800000,
        "extentTree": ExtentTree {
          "extents": [
            Extent {
              "length": 4,
              "start": 1,
            },
          ],
          "fileSize": 4,
        },
        "index": 1,
        "name": "test.txt",
        "size": 4,
        "state": "used",
        "type": "file",
      },
      "free": false,
      "reserved": false,
      "used": true,
    },
  },
  {
    "offset": 4,
    "state": {
      "color": "#f477a3",
      "data": Inode {
        "color": "#f477a3",
        "dateCreated": 1577836800000,
        "extentTree": ExtentTree {
          "extents": [
            Extent {
              "length": 4,
              "start": 1,
            },
          ],
          "fileSize": 4,
        },
        "index": 1,
        "name": "test.txt",
        "size": 4,
        "state": "used",
        "type": "file",
      },
      "free": false,
      "reserved": false,
      "used": true,
    },
  },
]
`;
